_posts/2019/2019-04-30-how-to-easily-manage-ssh-key-api-implementation.md

```diff
@@ -0,0 +1,182 @@
+---
+layout: post
+title: How to easily manage ssh keys - API implementation
+date: 2019-03-31
+header_image: public/ssh-key-management.jpg
+header_position: bottom
+header_overlay: true
+category: tech-stories
+tags: ["idoit", "ssh", "development", "tools"]
+authors: ["Carsten"]
+about_authors: ["cseeger"]
+---
+
+In the previous post [How to easily manage SSH keys](/blog/tech-stories/how-to-easily-manage-ssh-keys/) we described how to manage SSH keys.
+Now it's time to implement the backend, to be able to store, request and filter our ssh-keys and create `authorized_keys` files.
+
+If you've read my other blog posts, you know I'm a big fan of [Golang](https://golang.org/){:target="_blank"}.
+So it's no surprise that both, the API and the frontend are written in Golang.
+I've also written and do maintain a Golang [client library impementation](https://github.com/cseeger-epages/i-doit-go-api){:target="_blank"} for i-doit that will be used here.
+
+Before we start, this post does not touch the basics of implementing a REST API.
+It would be too long to read, so it just covers the specifics of our use case.
+If it's your first REST API implementation you may want to look at some tutorials and/or frameworks on the internet.
+I've also written a small [library](https://github.com/cseeger-epages/restfool-go){:target="_blank"} for this purpose where you can find some basic [examples](https://github.com/cseeger-epages/restfool-go/tree/master/examples){:target="_blank"} implementing a simple HTTPS GET handler.
+In our case all data payloads are provided by the user via HTTP(S) POST.
+
+Let's concentrate on the CMDB-specific parts.
+
+To not mess with authentication we use the authentication of the i-doit API, and just pass username and password through (or create a token).
+We now need to get the user id.
+The following simple trick helps us here:
+
+```
+  a, err := goidoit.NewLogin(URL, APIKEY, USERNAME, PASSWORD)
+  if err != nil {
+    // put your error handling here
+  }
+
+  data, _ := a.Version()
+  // the version contains also user information we can use
+  id, err := strconv.Atoi(data.Result[0]["login"].(map[string]interface{})["userid"].(string))
+  if err != nil {
+    // put your error handling here
+  }
+```
+To store our ssh keys we can use the following structure:
+
+```
+// SSHKey contains ssh key information
+type SSHKey struct {
+  ID      int    `json:"id"`
+  Key     string `json:"key"`
+  Label   string `json:"label"`
+  Primary string `json:"primary"`
+  Date    string `json:"date"`
+}
+```
+and get the SSH keys for the user:
+
+```
+
+  keyData, err := a.GetCat(id, "C__CATG__CUSTOM_FIELDS_SSHKEYS")
+  if err != nil {
+    // put your error handling here
+  }
+
+```
+
+If you never worked with i-doit before the `C__CATG__CUSTOM_FIELDS_SSHKEYS"` is a global category index used to find specific categories.
```

Comment:
Is this the correct index?

Reply:
yep its used in the first post `Constant:     C__CATG__CUSTOM_FIELDS_SSHKEYS`

Reply:
i removed the global before category